R/nbhd_ses_variables.R

```diff
@@ -0,0 +1,121 @@
+#' @title
+#' Obtain commonly used neighbourhood-level socioeconomic (SES) variables
+#'
+#' @description
+#' The `nbhd_ses_variables()` function merges SES variables based on dissemination area ID.
+#'
+#' @param dbcon (`DBIConnection`)\cr
+#' A database connection to any GEMINI database.
+#'
+#' @param dissemination_area\cr
+#' Version of dissemination_area chosen. Only 2016 or 2021 are valid inputs. 
+#' 
+#' @param cohort\cr
+#' User specified cohort
+#' 
+#' @return (`data.frame` | `data.table`)\cr
+#' This function returns a `data.table` merged with dissemination area ID and SES variables.
+#'
+#'
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' drv <- dbDriver("PostgreSQL")
+#' dbcon <- DBI::dbConnect(drv,
+#'                         dbname = "db",
+#'                         host = "domain_name.ca",
+#'                         port = 1234,
+#'                         user = getPass("Enter user:"),
+#'                         password = getPass("password"))
+#'
+#' nbhd_ses_table <- nbhd_ses_variables(dbcon, 2016, cohort)
+#' }
+#'
+
+nbhd_ses_variables <- function (dbcon, dissemination_area, cohort) {
+  if (!RPostgreSQL::isPostgresqlIdCurrent(dbcon)) {
+    stop("\n Please input a valid database connection")
+  }
+  
+  ## write a temp table to improve querying efficiency
+  DBI::dbSendQuery(dbcon,"Drop table if exists temp_data;")
+  DBI::dbWriteTable(dbcon, c("pg_temp","temp_data"), cohort[,.(genc_id)], row.names = F, overwrite = T)
+  
+  ## only da=2016, 2021 are allowed values
+  if (!dissemination_area %in% c(2016, 2021)) {
+    stop("Only 2016 and 2021 are valid versions of dissemination area")
+  }
+  
+  ## message to show invalid meanings of null values
+  message('Null values or 9 or 999 represent unknown or missing estimates')
```

Comment:
Document an unexpected case where income quintile=999 (income continuous=999999) have known DA in 2021 data. See codes below to replicate. Based on this result, they represent encounter with DA available but missing income. Will aggregate those to 'DA available but census/ON Marg missing'.

For 2016 data, income quintile=9 (income continuous=99999999) have all DA ending in 9999, we can safely believe those are unknown DA. Would aggregate those as 'Missing DA'.

```
source('/mnt/nfs/home/shijia/GitHub/Rgemini/R/utils.R')
source('/mnt/nfs/home/shijia/GitHub/Rgemini/R/neighborhood_ses.R')

drv <- dbDriver("PostgreSQL") 
dbcon <- DBI::dbConnect(drv,
  dbname = "report_db_v3_9_1", # change DB to load later
  host = "prime.smh.gemini-hpc.ca",
  port = 5432,
  user = "shijia",
  password = getPass::getPass("password")
)

### check 2021 data
var_tbl2 <- DBI::dbGetQuery(dbcon, 'select tmp.genc_id, l.da21uid, s.c21_vismin, s.c21_vismin_not, s.qnatippe, s.qnbtippe, s.qaatippe, s.qabtippe, s.atippe, s.btippe, s.c21_immsta, s.c21_immsta_imm, s.c21_ed_15over_postsec, s.c21_ed_15over, s.c21_ed_25to64_postsec, s.c21_ed_25to64, "households_dwellings_DA21", "material_resources_DA21", "age_labourforce_DA21", "racialized_NC_pop_DA21", "households_dwellings_q_DA21", "material_resources_q_DA21", "age_labourforce_q_DA21", "racialized_NC_pop_q_DA21"

                                          from temp_data tmp

                                          left join locality_variables l
                                            on l.genc_id = tmp.genc_id
                                          left join lookup_statcan_v2021 s
                                            on l.da21uid = s.da21uid; ') %>% as.data.table()
tst <- var_tbl2[!is.na(da21uid)]
tst[qnatippe==999]$da21uid %>% unique
 [1] 46230062 10090080 48150089 35510100 10090107 10090124 35390757 35370839 35600308 10090117 10090079 48040059
```

Reply:
It's unexpected in the sense that income=9 actually has valid DA in 2021 data. This affects my codes, but we are still using 2 types of missingness. See edits for details later

Reply:
Please re-review the information I shared above. What you describe refers to the 2nd type of missingness I mentioned - this is very much expected.

Your current code is still not correct. Why do you assume DA is invalid if `qnatippe == 9`, and why do you only make this assumption for 2016 data? As mentioned above, PCCF explicitly mentions the values referring to invalid DAs. Please use that in your code. The unique DAs you get with your code above are NOT invalid DAs. Again, they are valid DAs but with missing income data (see 2nd type of missingness).

Reply:
Hi @shijiaSMH: This is still not working as expected. Why are you applying this `if` statement when cleaning up missing values? `(census_year != 2016 & tnl != 1)`. This means that none of this code (including the warning messages) are executed when the user runs this function with `census_year = 2016` on the new DB version with 2 statcan tables. Please carefully check the function logic and test it with different user inputs to make this works as intended. 
In this context: I would also highly recommend removing some redundancy in the code, so it's easier to catch bugs like this.

Reply:
Hi @loffleraSMH, thanks for catching this. Now that I think of it, we have 2 options: 1) keep this message, not clean up 9s in income; 2) remove this message, clean up 9s in income.

Which seems better? 

It's quite strange about this one: (census_year != 2016 & tnl != 1). This means that none of this code (including the warning messages) are executed when the user runs this function with census_year = 2016 on the new DB version with 2 statcan tables.

I tested on both clean DB and report DB, things all run normally. Thanks for flagging, I'll test again.

Reply:
Hi @shijiaSMH: Please see above, we want to do both - 1) first, set 9s in income to `NA` since they represent missing values **AND** 2) show a warning message about the final % of `NA` (after setting 9s to `NA`).

If `census_year = 2016` and 2 tables are found (e.g., in `report_db_v3`) the `if` statement returns `FALSE`. Also, why would this `if` statement be relevant here at all? The clean-up/warning messages should be run regardless of the specific input arguments/DB version. Let's make sure the code logic is clear here.

Reply:
Hi @loffleraSMH, thanks again for those suggestions. I believe all those points are addressed in the update.

Note. I the function ran normally on `report_db_v3_9_1`, `drm_cleandb_v2_1_3`, `drm_cleandb_v3_0_0`. I conducted the tests 2 times. 

Please keep me posted if anything doesn't work from your end.